src/main/java/com/microsoft/azure/functions/worker/reflect/EnhancedClassLoaderProvider.java

```diff
@@ -19,12 +19,25 @@ public EnhancedClassLoaderProvider() {
      */
     @Override
     public ClassLoader createClassLoader() {
-        URL[] urlsForClassLoader = new URL[urls.size()];
-        urls.toArray(urlsForClassLoader);
+        return createURLClassLoaderInstance();
+    }
 
-        URLClassLoader classLoader = new URLClassLoader(urlsForClassLoader);
-        loadDrivers(classLoader);
-        return classLoader;
+    /**
+     * Create and return a singleton URL classloader
+     * @return instance of URLClassLoader
+     */
+    private URLClassLoader createURLClassLoaderInstance() {
+        if (classLoaderInstance == null) {
+            synchronized (lock) {
+                if (classLoaderInstance == null) {
+                    URL[] urlsForClassLoader = new URL[urls.size()];
+                    urls.toArray(urlsForClassLoader);
+                    classLoaderInstance = new URLClassLoader(urlsForClassLoader);
+                    loadDrivers(classLoaderInstance);
```

Comment:
good to assign the field at the very end, so that another thread can't come in and return `classLoaderInstance` prior to it being fully initialized
```suggestion
                    URLClassLoader loader = new URLClassLoader(urlsForClassLoader);
                    loadDrivers(loader);
                    classLoaderInstance  = loader ;
```